File: frontend/src/app/services/subscription.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {SubscriptionModel} from "../main-module/models/subscription.model";
import {Content} from "../main-module/models/content";

@Injectable({
    providedIn: 'root'
})
export class SubscriptionService {

    private _controllerUrlPrefix: string = '/api/subscriptions';

    constructor(private http: HttpClient) {
    }

    getSubscriptionsPaged(page: number, size: number): Observable<Content<SubscriptionModel>> {
        return this.http.get<Content<SubscriptionModel>>(this._controllerUrlPrefix + '?page=' + page + '&size=' + size);
    }

    getSubscriptionsByOwnerId(id: string): Observable<SubscriptionModel[]> {
        return this.http.get<SubscriptionModel[]>(this._controllerUrlPrefix + '/owners/' + id);
    }

    getSubscriptionByCategoryId(id: string, page: number, size: number): Observable<Content<SubscriptionModel>> {
        return this.http.get<Content<SubscriptionModel>>(this._controllerUrlPrefix + '/categories/' + id + '?page=' + page + '&size=' + size);
    }

    saveSubscription(subscription: SubscriptionModel): Observable<SubscriptionModel> {
        return this.http.post<SubscriptionModel>(this._controllerUrlPrefix, subscription);
    }

    getSubscriptionsByNameLike(name: string, page: number, size: number): Observable<Content<SubscriptionModel>> {
        return this.http.get<Content<SubscriptionModel>>(this._controllerUrlPrefix + '/search?name=' + name + "&page=" + page + "&size=" + size);
    }

    deleteSubscription(id: string): Observable<void> {
        return this.http.delete<void>(this._controllerUrlPrefix + '/' + id);
    }
}

File: apigateway/src/main/java/ru/otus/apigateway/service/api/SubscriptionService.java
package ru.otus.apigateway.service.api;

import ru.otus.apigateway.model.view.Content;
import ru.otus.apigateway.model.view.SubscriptionViewModel;

import java.util.List;

public interface SubscriptionService {

    Content<SubscriptionViewModel> findAll(int page, int size);

    Content<SubscriptionViewModel> findByNameLike(String name, int page, int size);

    Content<SubscriptionViewModel> findByCategoryId(Long id, int page, int size);

    List<SubscriptionViewModel> findByOwnerId(Long id);

    SubscriptionViewModel saveSubscription(SubscriptionViewModel subscription);

    void deleteSubscription(Long id);
}
